pkg/registry/helm.go

```diff
@@ -0,0 +1,363 @@
+func normalizeHelmURI(s string) (string, error) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return "", errors.Wrapf(err, "parsing Helm URI %q", s)
+	}
+
+	if _, ok := validHelmSchemes[u.Scheme]; !ok {
+		return "", errors.Errorf("%q is an invalid scheme for Helm repository", u.Scheme)
+	}
+
+	if strings.HasSuffix(u.Path, "index.yaml") {
+		return s, nil
+	}
+
+	u.Path = path.Join(u.Path, "index.yaml")
+	return u.String(), nil
+}
+
+// HelmRepository is metadata describing the contents of a Helm repository.
+type HelmRepository struct {
+	Charts map[string][]HelmRepositoryChart `json:"entries,omitempty"`
+}
+
+// Latest returns the latest version of charts in a Helm repository.
+func (hr *HelmRepository) Latest() []HelmRepositoryChart {
+	var out []HelmRepositoryChart
+
+	for _, chart := range hr.Charts {
+		if len(chart) > 0 {
+			sort.Slice(chart, func(i, j int) bool {
+				v1, err := semver.Make(chart[i].Version)
+				if err != nil {
+					return false
+				}
+				v2, err := semver.Make(chart[j].Version)
+				if err != nil {
+					return false
+				}
+
+				return v1.Compare(v2) == 1
+			})
+
+			out = append(out, chart[0])
+		}
+	}
+
+	return out
+}
+
+// HelmRepositoryChart is metadata describing a Helm Chart in a repository.
+type HelmRepositoryChart struct {
+	Description string   `json:"description,omitempty"`
+	Name        string   `json:"name,omitempty"`
+	URLs        []string `json:"urls,omitempty"`
+	Version     string   `json:"version,omitempty"`
+}
+
+type httpGet func(string) (*http.Response, error)
+
+type helmRepositoryClient struct {
+	url     string
+	httpGet httpGet
+}
+
+func newHelmRepositoryClient(u string, hg httpGet) (*helmRepositoryClient, error) {
+	u, err := normalizeHelmURI(u)
+	if err != nil {
+		return nil, errors.Wrap(err, "normalizing Helm repository URL")
+	}
+
+	if hg == nil {
+		hg = http.Get
+	}
+
+	return &helmRepositoryClient{
+		url:     u,
+		httpGet: hg,
+	}, nil
+}
+
+func (hrc *helmRepositoryClient) Repository() (*HelmRepository, error) {
+	resp, err := hrc.httpGet(hrc.url)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving repository index.yaml response")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("repository returned unexpected response %s", resp.Status)
+	}
+
+	defer func() {
+		cerr := resp.Body.Close()
+		if cerr != nil {
+			err = cerr
+		}
+	}()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading repository index.yaml body")
+	}
+
+	var hre HelmRepository
+
+	if err := yaml.Unmarshal(b, &hre); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling repository index.yaml")
+	}
+
+	return &hre, nil
+}
+
+func (hrc *helmRepositoryClient) Chart(name, version string) (*HelmRepositoryChart, error) {
+	repo, err := hrc.Repository()
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving repository")
+	}
+
+	if version == "" {
+		for _, chart := range repo.Latest() {
```

Comment:
This seems like a lot of extra work - we're sorting all versions of all charts, then doing a linear search for the desired chart name. Why instead not sort just `repo.Charts[name]`?